grpc/runtime/src/main/scala/io/buoyant/grpc/runtime/DecodingStream.scala

```diff
@@ -0,0 +1,288 @@
+ *
+ * Like other readable interfaces, [[DecodingStream.recv]] must be
+ * invoked serially, i.e. when there are no pending `recv()` calls on
+ * the stream.
+ */
+private[runtime] trait DecodingStream[+T] extends Stream[T] {
+  import DecodingStream._
+
+  protected[this] def frames: h2.Stream
+  protected[this] def decoder: ByteBuffer => T
+
+  // convenience
+  private[this]type Releasable = Stream.Releasable[T]
+
+  /**
+   * Holds HTTP/2 data that has not yet been returned from recv().
+   *
+   * These states hold a ByteBuffer-backed Buf, as well as a
+   * `Releaser`, which manages when HTTP/2 frames are released
+   * (i.e. for flow control).
+   */
+  private[this] var recvState: RecvState = RecvState.Empty
+
+  /**
+   * Ensures that at most one recv call is actively being processed,
+   * so that read-and-update access to `recvState` is safe.
+   */
+  private[this] val recvMu = new AsyncMutex()
+
+  /**
+   * Obtain the next gRPC message from the underlying H2 stream.
+   */
+  override def recv(): Future[Releasable] =
+    recvMu.acquireAndRun {
+      // Now we have exclusive access of `recvState` until the call
+      // completes. Start by trying to obtain a message directly from
+      // the buffer:
+      val updateF = decode(recvState, decoder) match {
+        case (s, Some(msg)) => Future.value(s -> Return(msg))
+        case (s, None) => read(s)
+      }
+      updateF.flatMap(_updateBuffer)
+    }
+
+  /**
+   * Update the receive buffer before returning the result
+   * (and i.e. releasing the mutex).
+   */
+  private[this] val _updateBuffer: ((RecvState, Try[Releasable])) => Future[Releasable] = {
+    case (rb, v) =>
+      recvState = rb
+      Future.const(v)
+  }
+
+  /**
+   * Read from the h2 stream until the next message is fully buffered.
+   */
+  private[this] def read(s0: RecvState): Future[(RecvState, Try[Stream.Releasable[T]])] = {
+    frames.read().transform {
+      case Throw(rst: h2.Reset) => Future.exception(Stream.Closed) // XXX FIXME
+      case Throw(e) => Future.exception(e)
+      case Return(t: h2.Frame.Trailers) => Future.exception(Stream.Closed) // XXX FIXME
+      case Return(f: h2.Frame.Data) =>
+        decodeFrame(s0, f, decoder) match {
+          case (s1, Some(msg)) => Future.value(s1 -> Return(msg))
+          case (s1, None) =>
+            if (f.isEnd) Future.value(s1 -> Throw(Stream.Closed)) // XXX FIXME
+            else read(s1)
+        }
+    }
+  }
+}
+
+object DecodingStream {
+
+  def apply[T](
+    req: h2.Request,
+    decodeF: ByteBuffer => T
+  ): Stream[T] = new DecodingStream[T] {
+    protected[this] val frames: h2.Stream = req.stream
+    protected[this] def decoder: ByteBuffer => T = decodeF
+  }
+
+  def apply[T](
+    rsp: h2.Response,
+    decodeF: ByteBuffer => T
+  ): Stream[T] = new DecodingStream[T] {
+    protected[this] val frames: h2.Stream = rsp.stream
+    protected[this] def decoder: ByteBuffer => T = decodeF
+  }
+
+  private sealed trait RecvState
```

Comment:
This sealed trait only has one subtype.  Are we planning to add more RecvState states?  Or should we just rename Buffer to RecvState and ditch the sealed trait?